ocrd/ocrd/mets_server.py

```diff
@@ -0,0 +1,253 @@
+        self.port = port
+        self.socket = socket
+        self.log = getLogger('ocrd.workspace_client')
+
+
+    def startup(self):
+
+        # XXX HACK 
+        # circumventing dependency injection like this is bad and
+        # needs to be refactored once it's all runnign
+        workspace = self.workspace
+
+        app = FastAPI(
+            title="OCR-D METS Server",
+            description="Providing simultaneous write-access to mets.xml for OCR-D",
+        )
+
+        @app.exception_handler(ValidationError)
+        async def exception_handler_invalid400(request: Request, exc: ValidationError):
+            return JSONResponse(status_code=400, content=exc.errors())
+
+        @app.exception_handler(FileExistsError)
+        async def exception_handler_invalid400(request: Request, exc: FileExistsError):
+            return JSONResponse(status_code=400, content=str(exc))
+
+        @app.exception_handler(re.error)
+        async def exception_handler_invalid400(request: Request, exc: re.error):
+            return JSONResponse(status_code=400, content=f'invalid regex: {exc}')
+
+        @app.get("/", response_model=OcrdFileListModel)
+        async def find_files(
+            file_grp : Union[str, None] = None,
+            file_id : Union[str, None] = None,
+            page_id : Union[str, None] = None,
+            mimetype : Union[str, None] = None,
+        ):
+            """
+            Find files in the mets
+            """
+            found = workspace.mets.find_all_files(fileGrp=file_grp, ID=file_id, pageId=page_id, mimetype=mimetype)
+            return OcrdFileListModel.create(found)
+
+        @app.put('/')
+        def save():
+            return workspace.save_mets()
+
+        @app.post('/', response_model=OcrdFileModel)
+        async def add_file(
+            data : bytes = File(),
+            file_grp : str = Form(),
+            file_id : str = Form(),
+            page_id : Union[str, None] = Form(),
+            mimetype : str = Form(),
+            url : str = Form(),
+        ):
+            """
+            Add a file
+            """
+            # Validate
+            file_resource = OcrdFileModel.create(file_grp=file_grp, file_id=file_id, page_id=page_id, mimetype=mimetype, url=url)
+            # Add to workspace
+            kwargs = file_resource.dict()
+            kwargs['page_id'] = page_id
+            kwargs['content'] = data
+            kwargs['local_filename'] = kwargs.pop('url')
+            workspace.add_file(**kwargs)
+            return file_resource
+
+        @app.get('/file_groups', response_model=OcrdFileGroupListModel)
+        async def file_groups():
+            return {'file_groups': workspace.mets.file_groups}
+
+        @app.post('/agent', response_model=OcrdAgentModel)
+        async def add_agent(agent : OcrdAgentModel):
+            kwargs = agent.dict()
+            workspace.mets.add_agent(**kwargs)
+            return agent
+
+        @app.delete('/')
```

Comment:
I personally don't like stopping the server with http delete. I would use the fastapi-stop-hook to save the workspace: https://fastapi.tiangolo.com/advanced/events/#shutdown-event.

Reply:
Good point about the shutdown event hook. But do you know any way of "properly" stopping uvicorn once it is started?

I like an explicit stop-server call because that allows for more automation, you could run the server in the background for as long as you need and then send `DELETE` and it's gone, without having to resort to `kill -TERM` the process or using shell job control.

Reply:
I don't know how to stop it properly. I found this www.github.com/tiangolo/fastapi/issues/2928 and skimmed it (very briefly, too much text) and I think there is nothing I would call properly. The more I think about the delete/stop thing the better I think it is so it would be ok for me this way now (so I set this thread to resolved).